### PR TITLE
kubeadm-operator: implement the task controller

### PR DESCRIPTION
**Pull request by fabriziopandini**
This PR implements the task controller

NB. I'm splitting operator implementation in a series of PRs to make things easier to review, but the backside of this is that there are some TODO's, that will be addressed in follow up PRs when all the operators will be in places.

If this can help. we can arrange for a shared PR review

/kind feature
/sig cluster-lifecycle
/area kubeadm-operator

/priority important-soon

/cc @neolit123
/cc @ereslibre
/cc @yagonobre
/cc @yastij

### CODE REVIEW

File: operator/controllers/runtimetask_controller.go [hunk 1]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+
+	// update status with paused override setting from top level objects
+	task.Status.Paused = pausedOverride
+}
+
+func (r *RuntimeTaskReconciler) reconcileNormal(executionMode operatorv1.OperationExecutionMode, task *operatorv1.RuntimeTask, log logr.Logger) error {
+	// If the Task doesn't have finalizer, add it.
+	//if !util.Contains(task.Finalizers, operatorv1.RuntimeTaskFinalizer) {
+	//	task.Finalizers = append(task.Finalizers, operatorv1.RuntimeTaskFinalizer)
+	//}
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: remove commented code or add TODO?

File: operator/controllers/runtimetask_controller.go [hunk 2]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+	patchHelper, err := patch.NewHelper(task, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always attempt to Patch the Task object and status after each reconciliation.
+	defer func() {
+		if err := patchHelper.Patch(ctx, task); err != nil {
+			log.Error(err, "failed to patch Task")
+			if rerr == nil {
+				rerr = err
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: trying to understand here:
if rerr is already non-nil, the error from Patch() will not be returned but only logged - is this OK?

File: operator/controllers/runtimetask_controller.go [hunk 3]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+	r.reconcilePauseOverride(operationPaused, task)
+
+	// Handle deleted Task
+	if !task.DeletionTimestamp.IsZero() {
+		err = r.reconcileDelete(task)
+		if err != nil {
+			return
+		}
+	}
+	// Handle non-deleted/non recovered Task
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: non-recovered

File: operator/controllers/runtimetask_controller.go [hunk 4]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+	// NB. in case or task recovered from error, we are forcing another reconcile before actually
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: or task -> of a task

File: operator/controllers/runtimetask_controller.go [hunk 5]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+	}
+
+	// Always reconcile Task Phase at the end
+	r.reconcilePhase(task)
+
+	return
+}
+
+func (r *RuntimeTaskReconciler) reconcileRecovery(executionMode operatorv1.OperationExecutionMode, task *operatorv1.RuntimeTask, log logr.Logger) bool {
+	// if there are no error, return
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: is no error
or 
are no errors

File: operator/controllers/runtimetask_controller.go [hunk 6]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+	if executionMode == operatorv1.OperationExecutionModeControlled {
+		task.Status.Paused = true
+	}
+	return nil
+}
+
+func (r *RuntimeTaskReconciler) reconcileDelete(task *operatorv1.RuntimeTask) error {
+
+	// Task is deleted so remove the finalizer.
+	//task.Finalizers = util.Filter(task.Finalizers, operatorv1.RuntimeTaskFinalizer)
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: add TODO or remove commented code?

File: operator/controllers/runtimetask_controller.go [hunk 7]
```diff
@@ -18,36 +18,324 @@ package controllers
...
+	if task.Status.Phase == "" {
+		task.Status.SetTypedPhase(operatorv1.RuntimeTaskPhasePending)
+	}
+
+	// Set the phase to "running" if start date is set.
+	if task.Status.StartTime != nil {
+		task.Status.SetTypedPhase(operatorv1.RuntimeTaskPhaseRunning)
+	}
+
+	// Set the phase to "paused" if paused set.
```
Comment (neolit123) [inline operator/controllers/runtimetask_controller.go]: set -> is set

File: operator/commands/factory.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		return runUpgradeKubectlAndKubelet(c.UpgradeKubeletAndKubeactl, log)
+	}
+
+	if c.Pass != nil {
+		return nil
+	}
+
+	if c.Fail != nil {
+		time.Sleep(5 * time.Second)
+		return errors.New("command fail failed")
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply operator/commands/factory.go]: Pass, Wait, Fail are dummy commands I use to test the workflow. They will go away ASAP
Sleep gives me time to catch up with what's going on

File: operator/commands/preflight.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+	operatorv1 "k8s.io/kubeadm/operator/api/v1alpha1"
+)
+
+func runPreflight(spec *operatorv1.PreflightCommandSpec, log logr.Logger) error {
+	cmd := newCmd("kubeadm", "version")
+
+	lines, err := cmd.RunAndCapture()
+	if err != nil {
+
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply operator/commands/preflight.go]: done. BTW this is only a command test, not the final preflight implementation

File: operator/commands/factory.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if c.Wait != nil {
+		time.Sleep(time.Duration(c.Wait.Seconds) * time.Second)
+		return nil
+	}
+
+	return errors.New("invalid Task.Spec.[]CommandDescriptor. There are no command implementation matching this spec")
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply operator/commands/factory.go]: done

File: operator/controllers/runtimetask_controller.go (hunk 1 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: Finalizers implementation is still pending il all the controllers, I don't have yet had time to figure out if they should go away or not... Until then I prefer to keep them into my way

File: operator/controllers/runtimetask_controller.go (hunk 2 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: rerr is the returned error of the outside func, err the error from patch command inside defer.

If there is no error from the outside procedure, then we return the patch error, if any
If there and error from the outside procedure, we return this (and the patch error, if any is only logged)

File: operator/controllers/runtimetask_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: done

File: operator/controllers/runtimetask_controller.go (hunk 4 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: done

File: operator/controllers/runtimetask_controller.go (hunk 5 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: done

File: operator/controllers/runtimetask_controller.go (hunk 6 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: see comment above about finalizers. still WIP

File: operator/controllers/runtimetask_controller.go (hunk 7 above)
Comment (fabriziopandini) [reply operator/controllers/runtimetask_controller.go]: done

**Comment by fabriziopandini** [comment]
@neolit123 thanks for the feedback! everything should be addressed or answers provided

File: operator/controllers/runtimetask_controller.go (hunk 2 above)
Comment (neolit123) [reply operator/controllers/runtimetask_controller.go]: ok

**Pull request merged**